students/Gabe_Lamberth/lesson09/mailroom5.py

```diff
@@ -1,25 +1,196 @@
+                        "3": self.mail.add_donor,
+                        "4": self.mail.send_letter,
+                        "q": self.end_menu}
+
+    def display_menu(self):
+        """
+        Function main drives the Donor program. It first prompts the user for a selection to pass to the
+        correlated method.The program will end when the user selects the quit method
+        """
+        print("""
+        Welcome to the MailRoom Donor Program!
+        What would you like to do?
+        Type 1 - To display donors'
+        Type 2 - To generate a report'
+        Type 3 - To add new donor'
+        Type 4 - To send letters to donors'
+        Type q - To end the program'
+        """)
+
+    def end_menu(self):
+        print("Shutting down program, Goodbye", end='\n')
+        sys.exit(0)
+
+    def run(self):
+        while True:
+            self.display_menu()
+            choice = input("Enter value here >> ")
+            action = self.choices.get(choice)
+
+            if action:
+                action()
+            else:
+                print(f'{choice} is not a valid choice')
+
+
+class MailRoom:
+
+    patrons = {'Emma Watson': [50.00, 150.00, 2000.00, 3000.00],
+               'Carolyn Wilson': [5000.25, 2500.00],
+               'Andres Dominguez': [5015.00, 200.50, 35.50]}
+
+    donors = Donor(patrons)
+
+    def show(self):
+        print('Showing list of patrons in the database:\n')
+        print(self.donors.keys())
+
+    def give_report(self):
+        header = ('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift')
+        print('{:25} |{:^15} |{:^15} |{:^15}'.format(*header))
+        print('-' * 75)
+        # Uses composition from Donor class
+        print(''.join({f"{k:<30} ${sum(v):^,.2f}  {len(v):^20}  ${sum(v)/len(v):,.2f}\n"
+                       for (k, v) in self.donors.items()}))
+        print('-- Report Finished --\n')
+
+        # Method to send letters
+    def send_email(self, name, amount):
+        print(f"Thank you {name} for your generous ${amount:,.2f} donation")
+        print('-- Email Sent --\n')
+
+    def add_donor(self):
+        # donor menu used to help validate choices
+        values = ['N', 'n', 'No', 'NO']
+
+        while True:
+            choice = input("""
+            Would you like to add a new donor, Yes or No?
+            Enter value here >>  """)
+            if choice in values:
+                break
+            else:
+                first, last = input("\nPlease enter the new Donor's First and Last name:  ").split()
+                name = f'{first} {last}'.title()
+
+            if name in self.donors.keys():
+                print("Donor's name already exists")
+                continue
+            else:
+                self.donors[name] = []
+                try:
+                    amount = float(input('Please enter a donation amount: '))
+                except ValueError:
+                    print("The donation must be numeric like 1234.00")
```

Comment:
So if it catches this exception the user will have to go all the way to the beginning.  Might it be better to just ask for a new donation amount in another while loop?